### DISCUSSION TITLE
The re-design of `config_src/`

### DISCUSSION THREAD

**Discussion by adcroft** [mom-ocean/MOM6#1286] [Ideas]
The current directory tree of compilable source code is
```
.
|-- config_src
|   |-- coupled_driver
|   |-- dynamic
|   |-- dynamic_symmetric
|   |-- external
|   |   |-- GFDL_ocean_BGC
|   |   `-- ODA_hooks
|   |-- ice_solo_driver
|   |-- mct_driver
|   |-- nuopc_driver
|   |-- solo_driver
|   `-- unit_drivers
|-- docs
|-- pkg
|   |-- CVMix-src
|   `-- GSW-Fortran
`-- src
```
where the some of the cirectories within `config_src/` are exclusive. e.g. you should only include either `dynamic/` or `dynamic_symmetric/` in the search path for the compiler. Similarly only one of the `*_driver/` directories should be used, and the different choices yield different forms of the model.

We wish to add directories containing a thin layer that translates between a MOM6-defined API and infrastructure libraries. We envision different directories corresponding to different versions of infrastructure, initially fms1 and fms2, and later someone can add one for esmf. On the dev call we just discussed that rather than add yet more directories to `config_src/` we should re-structure into something like:
```
.
|-- config_src
|   |--drivers
|   |   |-- coupled_driver
|   |   |-- ice_solo_driver
|   |   |-- mct_driver
|   |   |-- nuopc_driver
|   |   |-- solo_driver
|   |   `-- unit_drivers
|   |-- external
|   |   |-- GFDL_ocean_BGC
|   |   `-- ODA_hooks
|   |   |-- infrastructure
|   |   |   |-- FMS1
|   |   |   |-- FMS2
|   |   |   |-- ESMF
|   |   |   `-- nompi
|   |-- memory
|   |   |-- dynamic_nonsymmetric
|   |   `-- dynamic_symmetric
|-- docs
|-- pkg
|   |-- CVMix-src
|   `-- GSW-Fortran
`-- src
```
Better yet would be to group the infrastructure by operation.  Proposals welcome...

**Comment by Hallberg-NOAA** [comment]
I would prefer not to have infrastructure under external, but I would promote it up one level to be at the same level as drivers and memory, to avoid excessively long pathnames and to reflect that at least one infrastructure directory is needed, whereas it is not necessary to use any of the other directories under external.

**Comment by adcroft** [comment]
Ah! That wasn't my intent. Making trees by hand is harder than making the actual tree of directories! I meant like this:
```
.
|-- config_src
|   |-- drivers
|   |   |-- coupled_driver
|   |   |-- ice_solo_driver
|   |   |-- mct_driver
|   |   |-- nuopc_driver
|   |   |-- solo_driver
|   |   `-- unit_drivers
|   |-- external
|   |   |-- GFDL_ocean_BGC
|   |   `-- ODA_hooks
|   |-- infrastructure
|   |   |-- FMS1
|   |   |-- FMS2
|   |   |-- ESMF
|   |   `-- nompi
|   |-- memory
|   |   |-- dynamic_nonsymmetric
|   |   `-- dynamic_symmetric
|-- docs
|-- pkg
|   |-- CVMix-src
|   `-- GSW-Fortran
`-- src
```

**Comment by klindsay28** [comment]
Is the `driver` suffix on the subdirectories of `drivers` needed?
It feels redundant, except perhaps for `unit_drivers` (which could be renamed `unit_tests`).

**Comment by alperaltuntas** [comment]
This makes perfect sense. No objections. One minor suggestion I have is to drop the "_driver" suffixes, since they are now redundant, I guess

```
.
|-- config_src
|   |-- drivers
|   |   |-- coupled
|   |   |-- ice_solo
|   |   |-- mct
|   |   |-- nuopc
|   |   |-- solo
|   |   `-- unit
```

**Reply by alperaltuntas** [reply]
@klindsay28 beat me to it.

**Reply by Hallberg-NOAA** [reply]
Dropping the _driver suffix is a good suggestion.

**Reply by sanAkel** [reply]
👍

**Comment by marshallward** [comment]
Once we get something working, a potential phase 2:
```
.
|-- config_src
|   |-- drivers
|   |   |-- coupled
|   |   |-- ice_solo
|   |   |-- mct
|   |   |-- nuopc
|   |   |-- solo
|   |   `-- unit
|   |-- external
|   |   |-- GFDL_ocean_BGC
|   |   `-- ODA_hooks
|   |-- calendar   # Calendar, model time computation, astronomy
|   |    |-- fms
|   |    `-- esmf
|   |-- io  # Logging, File I/O, etc
|   |   |-- binary
|   |   |-- netcdf
|   |   |-- fms1
|   |   |-- fms2
|   |   `-- pio
|   |-- mpp
|   |   |-- serial   # No parallelization
|   |   |-- mpi      # Native MPI calls
|   |   |-- fms      # FMS parallelization (whatever it may be)
|   |-- tracer       # Generic tracer ops
|   |    `-- fms
|   |-- diagnostic   # diagnostic manager
|   |    `-- fms
|   |-- ???
|   |-- memory
|   |   |-- dynamic_nonsymmetric
|   |   `-- dynamic_symmetric
|-- docs
|-- pkg
|   |-- CVMix-src
|   `-- GSW-Fortran
`-- src
```

**Comment by marshallward** [comment]
Also, I feel obliged to mention that submodules were introduced to address this problem.

They are not necessary to achieve what we want, but they would allow us to compile the `src` content and produce binaries and .mod files independently of `config_src`.  It could also allow us to "hot swap" more easily between implementations.  Without submodules, there will be a strong compile-time link between `src` and `config_src` content.  Submodules replace this with a one-way dependency (`config_src` on `src` only).

One downside is that many compilers do not have a long history of support for them, many having only recently added them.

I don't think we need to pursue this, but wanted to at least make sure it was part of the discussion.

**Comment by adcroft** [comment]
1. Should `io`, `mpp`, `tracer`, `diagnostic` be grouped under `infrastructure/`
2. ??? should include `timers`  (cpu/perf clocks). Not sure where `time_interp` or `horiz_interp` belong?
3. Do we need a `misc` for everything else like `MersenneTwistor`, `constants`, ... ?
4. What is `tracer` ? Haven't seen it before...
5. Should terminal i/o (logging) be separate from binary files (netcdf)?

**Reply by marshallward** [reply]
Sorry, that wasn't meant to be an exhaustive list, just an indication that `infrastructure` could be replaced with a set of particular components.

(I just grepped for non-MOM modules and looked at the various subroutines and variables, including `generic_tracer_mod`).

**Comment by sanAkel** [comment]
All of this opens up lots of possibilities. Once all sorted out and ready to go!
1.  Whenever there is a PR, will it have to be tested for **ALL** `infrastructures`?
2. Let's say someone develops another _option_, say for diagnostics, whose burden would it be maintain it? The one who developed it or _volunteers_?

**Reply by marshallward** [reply]
1. When we discussed this internally, we proposed that GFDL would confirm that FMS2 works across all tests, but would commit to supporting and fixing FMS1 when issues become known.

   I think we'll still make our best effort to support as many frameworks as possible, however.  I would also expect other labs to test their own infrastructure(s) of choice.

2. I personally I think the contributor should be responsible for the content, and If the contributor isn't willing to either review the code or keep it up to date, then we may want to reconsider accepting it.

   At the least, we should have a designated maintainer in mind for all future contributions.

**Reply by sanAkel** [reply]
Sounds good! 

On 2, I would add that the `designated maintainer` also keep compiler flags also up to date, at least for some _recent_ version of intel and or GNU?

**Comment by Hallberg-NOAA** [comment]
I have some questions that I would like to raise about documenting the interfaces with this new structure for FMS2:

- Should we document the interfaces that will come from the config_src/infra/.../ code in the code that will stay in the MOM6/src/framework code, or can we just rely on having them documented in the (eventually numerous?) versions that will appear in the config_src/infra/.../ code?

- If the latter option (documentation in config_src/infra/...), how do we want to handle any divergences between the documentation of the various versions? (The compilers and appropriate testing should keep the interfaces themselves consistent.)

- Do we need to add explicit interfaces to the config_src/infra/.../*.F90 code to document every interface inherited from FMS and eventually the other infrastructures?

**Reply by Hallberg-NOAA** [reply]
Upon reflection, we think that (1) every API offered to MOM6 code needs an explicit implementation in MOM6 code (perhaps just a very thin wrapper) - there will be no pass-throughs from the infrastructure to anything used by the MOM6/src code, and (2) each of these implementations is the right place for the definitive documentation.  We will use diffs to keep the APIs in the various config_src/infra directories consistently documented.

**Comment by adcroft** [comment]
We've realized it is very hard to separate the io and communications from each other. When you combine IO with comms, and see what's left, almost everything is interconnected and the point of the fine grain directory structure is lost. The only things that look self contained are things like calendar, but even that is then used by time_interp. We are thus thinking we might stay with something more like in `config_src/infrastructure/FMS1` form, as in https://github.com/NOAA-GFDL/MOM6/discussions/1286#discussioncomment-275600, and not the "phase 2" with more granularity. Make sense?

**Comment by Hallberg-NOAA** [comment]
After reflecting on various proposals and considering how they might actually be implemented I think that we are approaching a consensus for the new structure.  The new proposal is for the new structure to be the following:

```
|-- config_src
|   |--drivers
|   |   |-- FMS_cap
|   |   |-- ice_solo_driver
|   |   |-- mct_cap
|   |   |-- nuopc_cap
|   |   |-- (python_cap?)
|   |   |-- solo_driver
|   |   `-- unit_drivers
|   |-- external
|   |   |-- GFDL_ocean_BGC
|   |   `-- ODA_hooks
|   |-- infra
|   |   |-- FMS1
|   |   |-- FMS2
|   |   |-- (ESMF?)
|   |   `-- (nompi?)
|   `-- memory
|       |-- dynamic_nonsymmetric
|       `-- dynamic_symmetric
|-- docs
|-- pkg
|   |-- CVMix-src
|   `-- GSW-Fortran
`-- src
    |-- ALE
    |-- core
    |-- diagnostics
    |-- eqn_of_state
    |-- framework
    |-- initialization
    |-- ocean_data_assim
    |-- parameterizations
    |-- tracer
    `-- user
```

The directories in parentheses do not exist, but are just placeholders for things that someone might want to add later.
Under drivers, a cap is something that acts to connect MOM6 with a coupler or external driving software (e.g., python), whereas a driver is code with its own Fortran main().  The contents of the current `config_src/coupled_driver` would be moved to `config_src/drivers/FMS_cap`, reflecting the fact that this is driven by the GFDL / FMS coupler code.

  We are likely to start implementing this for a PR by mid-February, so any further comments by then would be appreciated.

**Reply by alperaltuntas** [reply]
The initial idea of having more granularity would enable partial and/or incremental adoption of a particular infrastructure. For instance, as a first step, NCAR could implement it's own I/O infrastructure and use FMS for everything else initially. Does this new structure allow for such incremental approach, or would we need to have a complete infrastructure replacement  outright?

**Reply by Hallberg-NOAA** [reply]
We looked into using a finer granularity, but a number of different infrastructure services (e.g,. I/O, cross-processor communication, error handling, time-types, grid-interpolation, etc.) use code that is sufficiently intertwined (at least with FMS) that this did not seem to be practical at this time.
I would also note that the challenge of coming up with a proper testing strategy for multiple infrastructures will be hard enough, let alone devising testing for, say, 3^4=81 permutations of the infrastructure with between 4 and 7 drivers.

**Reply by alperaltuntas** [reply]
So, the answer is no? (The question was "Does this new structure allow for incremental substitution?") I imagine one particular way of having an incremental substitution could be, perhaps, for NCAR to start maintaining its own FMS fork and incrementally replace parts of it. Since the MOM6 community is now committed to support multiple infrastructures, I think it would be viable, right?

**Reply by Hallberg-NOAA** [reply]
If the incremental improvements within the infrastructure code do not change any interfaces, we would not need a separate directory under infra, but if (like FMS1 and FMS2) the interfaces do change, we would obviously need to add a separate directory under infra to accommodate the different interfaces.  The idea here is that, yes, we could add the directories under infra to enable the use of a manageably small number of infrastructures that provide comparable services with different interfaces and routine names. I could see what you are discussing, @alperaltuntas, as possibly fitting into one of these two cases, but I expect that the answer to your broader question is "Yes".

**Reply by alperaltuntas** [reply]
Great, thanks!

**Reply by sanAkel** [reply]
@alperaltuntas do you happen to have pseudo code for what you have in mind? may be a gist if it is _long_?

**Reply by alperaltuntas** [reply]
@sanAkel, We don't have a concrete plan yet, but, as a first step, I am imagining a lightweight FMS patch to implement few CESM-specific functionalities like PIO interface and a CESM/ESMF-integrated global constants, logging/stdout implementations, etc. @Hallberg-NOAA's second characterization seems a more likely path (i.e., interface changes and so a new directory under infra).

**Reply by sanAkel** [reply]
Thank you @alperaltuntas. If possible, may be we could chat _offline_. My thoughts are bit sketchy right now, but we could follow your approach for 
>CESM/ESMF-integrated global constants, logging/stdout implementations,

of course with GEOS-MAPL constants, for instance, see https://github.com/GEOS-ESM/MAPL/blob/main/shared/MAPL_Constants.F90 as a _variant_ from our side.

I can only imagine that the ~~infra~~ interface change**s** would be more stronger with a diverse pool! 🌞

**Reply by alperaltuntas** [reply]
@sanAkel Sure! let's talk and identify any common goals/requirements and address them in coordination wherever possible.
